shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -1021,6 +1021,50 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
         val desug = If(IfThen(lhs, rhs), S(Var("false")))
         term.desugaredTerm = S(desug)
         typeTerm(desug)
+      case App(f: Term, a @ Tup(fields)) if (fields.exists(x => x._1.isDefined)) =>
+        def getLowerBoundFunctionType(t: SimpleType): List[FunctionType] = t.unwrapProvs match {
+          case PolymorphicType(_, AliasOf(fun_ty @ FunctionType(_, _))) =>
+            List(fun_ty)
+          case tt @ FunctionType(_, _) =>
+            List(tt)
+          case tv: TypeVariable =>
+            tv.lowerBounds.map(getLowerBoundFunctionType(_)).flatten
+          case ct @ ComposedType(pol, lhs, rhs) =>
+            if (pol === false) {
+              getLowerBoundFunctionType(lhs) ++ getLowerBoundFunctionType(rhs)
+            } else 
+              Nil
+          case _ =>
+            Nil
+        }
+        val f_ty = typeTerm(f)
+        val fun_tys: List[FunctionType] = getLowerBoundFunctionType(f_ty)
+
+        fun_tys match {
+          case fun_ty :: Nil =>
+            val hasUntypedArg = fun_ty match {
+              case FunctionType(TupleType(fields), _) =>
+                fields.exists(_._1.isEmpty)
+              case _ => die // FIXME happens when lhs of function type is not TupleType(possible?)
```

Comment:
Please don't crash for no reason here. Just return the safe option that makes sense, which is `false`, since this is meant as a conservative test (it's not too bad if we reject more than strictly needed).

This case actually _can_ happen, although it's hard to trigger.